master/buildbot/worker/latent.py

```diff
@@ -183,27 +209,26 @@ def _substantiation_failed(self, failure):
         return self._mail_missing_message(subject, text)
 
     def canStartBuild(self):
-        if self.insubstantiating:
+        if self.insubstantiating or self.retry_timer:
             return False
         return AbstractWorker.canStartBuild(self)
 
     def buildStarted(self, sb):
+        assert sb.isBusy()
         self._clearBuildWaitTimer()
 
     def buildFinished(self, sb):
-        AbstractWorker.buildFinished(self, sb)
-
+        assert not sb.isBusy()
         if not self.building:
             if self.build_wait_timeout == 0:
-                d = self.insubstantiate()
-                # try starting builds for this worker after insubstantiating;
-                # this will cause the worker to re-substantiate immediately if
-                # there are pending build requests.
-                d.addCallback(lambda _:
-                              self.botmaster.maybeStartBuildsForWorker(self.workername))
+                self.insubstantiate()
```

Comment:
The state DETACHED is for the workerforbuilder. Its weird that the state is managed here, as actually all the wfd for the same worker are supposed to be in the same state.

So actually insubstanciate will not move the wbd to detached state. The  worker will not take new build during insubstanciation because canStartBuild will return false whel self.insubstanciating = True

The wfb state is DETACHED during substantiation, though. I can't remember exactly the reason why I had to do that.

This is part of the cleanups that needs to be done once this thing is fully tested at integration level.

I think there are still lots of simplification of the state.

notably I think we shouldn't have a state for wfb. They are several boolean state which makes sense, and combine into a canStartBuild()

the new canStartBuild for a latent wfb would be:

if self.running -> false
if self.worker.deleting -> false
if self.worker.creating -> true
else -> true

the new canStartBuild for a normal wfb would be:

if self.running -> false
if not self.attached -> false
else -> true

what do you think @seankelly , @tomprince ?